core/dnsserver/onstartup.go

```diff
@@ -2,9 +2,16 @@ package dnsserver
 
 import (
 	"fmt"
+	"regexp"
 	"sort"
 )
 
+// checkDomainName() returns true if the given domain name follows RFC1035 preferred syntax.
+func checkDomainName(domainName string) bool {
+	matchDomain, _ := regexp.Compile("[A-Za-z0-9-]{1,}\\.[A-Za-z0-9-]{1,}\\.$)|^\\.$")
```

Comment:
Suggestion: It would be easier to first check for the root domain case separately than to figure out how to do it in regex.  Also, it would result in a shorter and easier to read regex.

Also, There is no benefit to compiling the regex in a separate step since it's only being evaluated once.  You could instead use single line [regexp.MatchString](https://pkg.go.dev/regexp#MatchString).  To avoid re-compiling the regex for every zone, the compile step needs to be done outside of the zone keys loop in `startUpZones`.

Reply:
I think so, there is no benefit to compile the regex but if I use [regexp.MatchString](https://pkg.go.dev/regexp#MatchString), code will adjust in two steps itself like now, first step to handle return values from `regexp.MatchString()` like `bool` and `err` and then result will be returned in the next step so I think this is good to go If you feel any issue with it so let me know I will update it. 

and your suggestion to check root-domain separately looks good and will be easier for anyone to understand regex, I will update the regex again to make it better.

Reply:
> ... I think this is good to go ...

I think the compile step should be done outside of the zone keys loop in startUpZones.

Reply:
As you suggested earlier
> Each zone key in `zones` needs to be validated.

So, I think according to that suggestion It's perfect but If you want me to edit it I'll do it.

Reply:
As we know `zones` is an Array of Strings, where right now only one string is present (That's root domain-name `.`), But in future when codebase will be deployed , There will be more than one domain-name in `zones` so for that we will need to
call `checkDomainName()` in a for-loop so that we can validate each `zone` of `zones` and now, we are doing the same if I call `checkDomainName()` outside of zone keys loop I'll have to create another for-loop in `startUpZones`.
Let me know if I'm wrong here or you have different vision for it. I'll enhance code as per your suggestion.

Reply:
@MdSahil-oss, 

> As we know `zones` is an Array of Strings ...

This is incorrect. `zones` is a map of `string` to `Config`, not an array.  `keys` is a [slice](https://www.godesignpatterns.com/2014/05/arrays-vs-slices.html) of domain names derived from the keys of the `zones` map.

> where right now only one string is present

That's not correct.  It can hold any number of strings.

It is inefficient to re-compile the regex for each domain string in the list.  It should be compiled once, and then the resulting compiled regex can be used to evaluate each domain.  It's not a major problem, since this only happens at startup, and the list of domain names will usually be short.  But IMO it's a good habit to be efficient in code when doing so is easy. Since it's easy to do here, it should be done.

>  I'll have to create another for-loop in startUpZones

No, you don't need to do that.

Please test your code using the Corefile below.  Put temporary print statements in the code (or use a debugger if you already have set up) to look at the value of `keys` in `startupZones`.

This is one Corefile, not 3.
```
# these zones are ok
example.com abc-123.com abc-123 abc.123-xyz a----3 {
 log
}

# these zones should display warnings
abc-^%$.example.com abc-.example.com 123-abc.example.com {
 log
}

# this zone is ok
. {
  log
}
```

Reply:
They fail because your regex requires at least two labels to be present.

Reply:
@chrisohaver should I improve Regex for single label either ?

Reply:
A single label is permitted in the preferred syntax. So yes.

Reply:
Okay... so I'm improving my Regex for single label too but don't forget there will be a dot `.` required as trailing characters for each `zone`. e.g
# these zones are ok
example.com. abc-123.com. abc-123. abc.123-xyz. a----3. {
 log
}

# these zones should display warnings
abc-^%$.example.com. abc-.example.com. 123-abc.example.com. {
 log
}

# this zone is ok
. {
  log
}

Reply:
Ok. Feel free to use the regex I provided earlier. You’ll still need to check for the root domain separately, but I think its ok to do that.